### PR TITLE
feat(worker): accept component info on data plane (#1527)

### LINKED ISSUE #1527: [FEATURE] Component version, release, and wire-protocol compatibility management
**Is your feature request related to a problem? Please describe.**
Curvine components do not have a single, traceable version contract:

- Cargo workspace version stays at `0.2.0` while the repository advances past `v0.4.0-alpha`, so local builds cannot be mapped back to a commit or tag.
- Worker heartbeat only reports a display string (`software_version`), which is not reliable enough for compatibility checks.
- Client-master has no handshake that returns master version, protocol version, minimum supported version, or capabilities.
- Client-worker data-plane requests do not carry component or protocol version information.
- Known incompatible versions have no code-default policy, compatibility matrix, or blocklist; operators are forced to manage compatibility manually.
- Release artifacts (binaries, images, Helm, Java/Python SDKs) derive versions from different places and drift from the release tag.

**Describe the solution you'd like**
Implement a version and compatibility design that covers all component versions, the three main protocol planes, and the release pipeline.

### Version model

Keep four independent version dimensions:

```text
release_version   X.Y.Z[-pre]        release/package version
git_version       commit/tag/branch  build provenance
protocol_version  u32                wire protocol version
schema_version    u64                storage/local state version
```

Git tag is the release anchor:

```text
Git tag:            vX.Y.Z[-alpha|-beta|-rc]
Cargo workspace:    equals X.Y.Z at release; next-dev after release
Local build:        derived from git describe; CI overrides with BUILD_VERSION
Images/Helm/SDK:    version from release tag with v removed
```

All components expose structured version info through `--version`, `--version-json`, and startup logs.

### Public version protocol

Add common protobuf messages:

```proto
message ComponentInfoProto {
  optional string component = 1 [default = ""];
  optional string release_version = 2 [default = ""];
  optional string git_commit = 3 [default = ""];
  optional string git_tag = 4 [default = ""];
  optional string git_branch = 5 [default = ""];
  optional uint32 protocol_version = 6 [default = 1];
  optional uint32 min_protocol_version = 7 [default = 1];
  repeated string capabilities = 8;
}

enum CompatibilityModeProto {
  DIAGNOSE = 1;
  ENFORCE = 2;
}

message ServerCompatibilityInfoProto {
  required ComponentInfoProto server = 1;
  required string min_worker_version = 2 [default = ""];
  required string min_client_version = 3 [default = ""];
  required CompatibilityModeProto compatibility_mode = 4 [default = DIAGNOSE];
  repeated string blocked_versions = 5;
}
```

Proto field ID policy:

- `1-999`: message business fields.
- `1000-1999`: cross-cutting protocol metadata such as version, compatibility, capabilities, trace.
- `2000+`: reserved for future infrastructure or protocol extension.
- New messages use low sequential field numbers; protobuf does not allow field number `0`.

### Protocol planes

| Protocol plane | Direction | Typical RPC | Version constant |
| --- | --- | --- | --- |
| FS/Control | Client -> Master | Mkdir, CreateFile, GetMasterInfo, Mount, SubmitTransfer | `FS_PROTOCOL_VERSION` |
| Data | Client -> Worker | WriteBlock, ReadBlock, WriteBlocksBatch, FilesBatchWrite | `DATA_PROTOCOL_VERSION` |
| Cluster | Worker -> Master | WorkerHeartbeat, WorkerBlockReport | `WORKER_PROTOCOL_VERSION` |
| Transfer/Job | Client/Master/Worker | SubmitTransfer, TransferTaskReport, SubmitTask | `TRANSFER_PROTOCOL_VERSION` |
| Replication | Worker/Master | SubmitBlockReplicationJob, ReportBlockReplicationResult | `REPLICATION_PROTOCOL_VERSION` |

Implementation approach:

- Client-master: reuse `GetMasterInfo`; add `component_info` and `compatibility` fields and perform a handshake during client initialization.
- Worker-master: keep `software_version`; add structured `component_info` to heartbeat and store it on the master.
- Client-worker: add optional `component_info` to data-plane requests; worker caches peer version per connection.
- Keep the fixed RPC frame unchanged and place protocol versions in application protobuf fields for backward compatibility.

### Compatibility policy

Compatibility is decided by code and protocol contracts, not by operator configuration alone:

- Protocol check: `server.min_protocol_version <= peer.protocol_version <= server.protocol_version`.
- Component check: master-worker and master-client version ranges.
- Capability check: a feature is enabled only when both peers declare the capability.
- `diagnose` mode records warnings and allows requests; `enforce` mode rejects incompatible peers.
- `blocked_versions` is an emergency operator override, not the normal compatibility mechanism.
- Known incompatible versions are handled by bumping protocol versions, publishing patch releases that tighten defaults, and recording unsupported pairs in a compatibility matrix and release notes.

### Release management

```text
1. Release PR updates Cargo workspace version, Helm appVersion, and SDK packaging versions.
2. Tag the exact release commit as vX.Y.Z.
3. CI validates tag == Cargo version and injects BUILD_VERSION.
4. Binaries, images, Helm, Java, Python, and release notes use the same version.
5. After release, main bumps to the next development version.
```

Also fix `build/build.sh` so it reads the workspace package version correctly and support `BUILD_VERSION` override in `build.rs`.

**Describe alternatives you've considered**
- Compare raw git version strings for compatibility: brittle and cannot express ranges, protocol versions, or capabilities.
- Make compatibility config-only: turns a product contract into per-cluster manual operations and blocks safe rolling upgrades.
- Change the fixed RPC frame now: would break all legacy components and is not required by this design.
- Keep `CARGO_PKG_VERSION` as the only source: makes local builds unidentifiable and keeps release drift unresolved.

**Additional context**
This design extends issue #588 and references prior work in closed PR #595. It is also related to #869 because version compatibility is a prerequisite for smooth cluster upgrades.

# Goal

| Sub-requirement | Problem solved |
| --- | --- |
| Unified component version output | All binaries, FUSE, CLI, CSI, and SDKs can report the same structured provenance. |
| Git tag/Cargo/release alignment | Local builds and release artifacts map back to one commit and tag. |
| Client-master protocol handshake | Master can return version, protocol, and capability information before client work starts. |
| Worker-master structured version reporting | Master can validate and display worker versions instead of parsing display strings. |
| Client-worker protocol negotiation | Data-plane connections can detect peer versions and capabilities independently. |
| Compatibility matrix and known-incompatible handling | Mixed-version clusters can be diagnosed, rolled out, and enforced with clear policy. |
| Release pipeline validation | Binaries, images, Helm, and SDK packages are built from one version source. |

# Not Goal

- Changing the fixed RPC frame layout.
- Implementing metadata/schema migration semantics beyond version checks and migration entry points.
- Defining new SDK APIs or language-specific compatibility behavior.
- Solving Kubernetes pod placement or in-place upgrade mechanics covered by #869.

# Test Plan

| Test case | Test objective | Modules involved |
| --- | --- | --- |
| Version parsing and git describe | Local and tag builds produce stable, traceable versions | build.rs, build.sh, curvine-sys |
| Master-worker mixed versions | New master handles old and new workers in diagnose/enforce modes | master, worker, proto |
| Client-master handshake | Client caches master version/protocol/capabilities; legacy master degrades | client, master, proto |
| Client-worker data plane | Worker identifies peer version per connection and rejects incompatible protocol | client, worker, proto |
| Release validation | Tag, Cargo, Helm, Java/Python, and image versions match | workflows, packaging |
| Compatibility matrix | Documented version pairs and blocked versions behave as expected | docs, config, checker |
| E2E mixed cluster | FUSE, CLI, Java SDK, and Python SDK can connect and read/write in a mixed cluster | client, worker, master, SDKs |

Implementation can be split into milestones:

- M1: version module and build pipeline.
- M2: public protocol and client-master handshake.
- M3: worker-master structured version reporting.
- M4: client-worker protocol negotiation.
- M5: release and SDK version alignment.
- M6: compatibility matrix, tests, and documentation.

### PR DESCRIPTION
**Pull request by lzjqsdd**
# Summary

Add an optional `component_info = 1000` field to the client-worker data-plane requests (`BlockWriteRequest`, `BlockReadRequest`, `BlocksBatchWriteRequest`, `BlocksBatchCommitRequest`, `FilesBatchWriteRequest`), reusing the existing `ComponentInfoProto`. The client attaches its structured version to every open/commit frame it sends to workers, and the worker records the peer's component info per connection (first request that carries it wins; a mid-connection change is logged and re-recorded). This is the T9 step of the component version-management plan (#1527): workers can now identify the version/protocol of the client on each data connection without breaking old components.

# Issue Describe / Design

Closes #1527

Design decisions (from the version-management design):
- Reuse the reserved `1000+` cross-cutting field range so existing business fields (1-999) stay untouched; legacy workers ignore the unknown field via standard protobuf semantics.
- The worker only records the peer version at the connection level (`WorkerHandler` is created per connection), leaving protocol negotiation / diagnose-enforce checks to the follow-up client-worker compatibility layer (T10).
- Legacy clients that omit `component_info` keep working: the worker treats absence as a legacy/unknown peer, records nothing and never rejects.
- Running (data-carrying) frames use `DataHeaderProto` and carry no version payload, so the client reports on open/commit frames and the worker caches on the first request.

# Changes

| Module / File | Change | Impact on existing behavior |
| ------------- | ------ | --------------------------- |
| `crates/common/curvine-proto/proto/worker.proto` | Add optional `component_info = 1000` to `BlockWriteRequest`, `BlockReadRequest`, `BlocksBatchWriteRequest`, `BlocksBatchCommitRequest`, `FilesBatchWriteRequest` | None (new optional field, old components ignore it) |
| `crates/client/curvine-client-core/src/block/block_client.rs` | Build the client's `component_info` once per `BlockClient` and attach it to all data-plane open/commit requests | None (new field only) |
| `curvine-worker/src/worker/handler/worker_handler.rs` | Connection-level peer cache: extract `component_info` from data-plane frames, record on first sight, expose `peer_component_info()` | None (record-only, diagnose by default) |
| `curvine-worker/src/worker/handler/batch_write_handler.rs` | Propagate `component_info` from the batch-open frame into the per-block internal `BlockWriteRequest` | None |
| `curvine-worker/src/worker/worker_server.rs` | Initialize the per-connection peer cache | None |
| `curvine-worker/Cargo.toml` | Add `prost` dev-dependency for handler unit tests | None |
| `crates/common/curvine-proto/tests/worker_dataplane_compat_test.rs` | New wire-level compat tests: round-trip with `component_info`, legacy client decode (old wire), unknown high-field skipping | Test only |
| `curvine-server/tests/worker_test.rs`, `spdk_stress_test.rs`, `spdk_worker_test.rs`, `crates/common/curvine-proto/tests/proto_pipeline_test.rs` | Set `component_info: None` in test request constructions | Test only |

# Test verified

| Test case | Result | Notes |
| --------- | ------ | ----- |
| `cargo test -p curvine-proto` (incl. new `worker_dataplane_compat_test`) | PASS | 30 passed |
| `cargo test -p curvine-client-core` | PASS | 35 passed |
| `cargo test -p curvine-worker --lib` (incl. new handler peer-cache tests) | PASS | 18 passed, 1 ignored |
| `cargo test -p curvine-model --lib` | PASS | 36 passed |
| `cargo check --workspace --all-targets` | PASS | |
| `make format` (fmt + clippy `--deny warnings --all-targets`) | PASS | exit 0 |

Coverage for the T9 acceptance criteria:
- New/old client requests: legacy wire decode tests for all five messages (old client -> new worker), plus round-trip tests with `component_info` (new client -> new worker).
- Connection-level peer cache: worker handler unit tests covering extraction from each data-plane frame, running frames carrying none, legacy/unknown peers, and first/unchanged/changed cache transitions.

# Dependencies

- Related PRs: #1585 (T8 compat checker), #1584 (T7 worker heartbeat version), #1582 (T6 client-master handshake), #1581 (T5 GetMasterInfo extension)
- Related issues: #1527
- Blocks / blocked by: blocked by T4 (proto messages, already merged); T10 builds on this PR

### CODE REVIEW

File: crates/client/curvine-client-core/src/block/block_client.rs [hunk 1]
```diff
@@ -422,6 +436,7 @@ impl BlockClient {
             files: file_data,
             req_id,
             seq_id,
+            component_info: Some(self.component_info.clone()),
```
[the remark this answers is not in the document]
Comment (lzjqsdd) [reply crates/client/curvine-client-core/src/block/block_client.rs]: Thanks. Fixed in 64c05df4: the client now omits `component_info` on the `FilesBatchWriteRequest` running frame (batch open/commit frames keep it), keeping the content-bearing header lean and avoiding worker-side large-header decoding.

File: curvine-worker/src/worker/handler/worker_handler.rs [hunk 2]
```diff
@@ -85,7 +98,113 @@ impl MessageHandler for WorkerHandler {
...
+                .and_then(|req| req.component_info),
+            (RpcCode::WriteBlocksBatch, RequestStatus::Running) => msg
+                .parse_header::<FilesBatchWriteRequest>()
+                .ok()
+                .and_then(|req| req.component_info),
+            (RpcCode::WriteBlocksBatch, RequestStatus::Complete | RequestStatus::Cancel) => msg
+                .parse_header::<BlocksBatchCommitRequest>()
+                .ok()
+                .and_then(|req| req.component_info),
+            _ => None,
```
[the remark this answers is not in the document]
Comment (lzjqsdd) [reply curvine-worker/src/worker/handler/worker_handler.rs:176]: Thanks. Fixed in 64c05df4: `extract_component_info` no longer parses running frames — the `(WriteBlocksBatch, Running)` branch (FilesBatchWriteRequest) is removed, so the hot path never decodes the large content-bearing header. Peer metadata is recorded from the batch open/commit frames only, which always carry `component_info` from the new client.

**Comment by lzjqsdd** [comment]
Addressed both suppressed suggestions from the Copilot reviews in 2c621c0e.

**1. worker_handler.rs — repeated header decode on the data-plane control path**

`record_peer_component_info()` used to call `extract_component_info()` on every open/commit/cancel frame, decoding the same header the business handlers decode again (e.g. `WriteContext::from_req`, `BatchWriteHandler::open_batch`). The worker now resolves the connection peer **exactly once**, from the first data-plane **open** frame:

- new client carries `component_info` on it → `Known`;
- legacy client does not → `Legacy`.

After resolution (either state) the worker never parses another header for peer metadata, so at most one frame per connection pays the decode cost. Running (data-carrying) frames and commit frames never resolve, and non-data-plane RPCs (task/transfer) are excluded from resolution. The `Option<ComponentInfoProto>` cache was replaced by a `ConnectionPeer { Unknown | Legacy | Known }` enum; the first/unchanged/changed bookkeeping that was unreachable under record-once semantics was dropped.

**2. batch_write_handler.rs — inner `header` shadowing**

The inner `BlockWriteRequest` binding in `open_batch` is renamed to `block_header`, so it no longer shadows the outer `BlocksBatchWriteRequest` header and the `component_info` propagation reads unambiguously.

Tests: the cache-state test was replaced with resolution tests covering new/legacy open frames, running/commit frames (never resolve) and non-data-plane RPCs (never resolve). `make format` (fmt + clippy `--deny warnings --all-targets`) passes; `cargo test -p curvine-worker --lib` passes (21 passed, 1 ignored).

**Review by szbr00 — APPROVED** [review]
